Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,266 @@
+
+                    unsigned numVertices = layer.geometry.numberOfVertices();
+                    for (unsigned i = 0; i < numVertices; i++) {
+                        auto v = toLayerTransform->mapPoint(layer.geometry.vertexAt(i));
+                        clipVertices.append(v.x());
+                        clipVertices.append(v.y());
+                    }
+                }
             }
+        });
 
-            ClipPath clipPath(WTFMove(points), clipBuffer->bufferID(), polygon.clipVertexBufferOffset * sizeof(float));
+        unsigned clipBufferSize = clipVertices.size() * sizeof(float);
+        auto clipBuffer = textureMapper.acquireBufferFromPool(clipBufferSize, TextureMapperGPUBuffer::Type::Vertex);
+        clipBuffer->updateData(clipVertices.data(), 0, clipBufferSize);
 
-            paintLayerFunction(polygon.layer, clipPath);
-        }
+        // Paint
+        traverseTree(*root, [&clipVertices, &clipBuffer, &paintLayerFunction](LayerNode& node) {
+            for (auto& layer : node.layers) {
+                unsigned numberOfClipVertices = layer.isSplitted ? layer.geometry.numberOfVertices() : 0;
+
+                Vector<FloatPoint> points;
+                if (numberOfClipVertices > 0) {
+                    points.reserveCapacity(numberOfClipVertices);
+                    auto xy = clipVertices.subvector(layer.clipVertexBufferOffset, numberOfClipVertices * 2);
+                    for (size_t i = 0; i < xy.size(); i += 2)
+                        points.append(FloatPoint(xy.at(i), xy.at(i + 1)));
+                }
+
+                ClipPath clipPath(WTFMove(points), clipBuffer->bufferID(), layer.clipVertexBufferOffset * sizeof(float));
+
+                paintLayerFunction(layer.textureMapperLayer, clipPath);
+            }
+        });
+    }
+}
+
+TextureMapperLayer3DRenderingContext::SweepAndPrunePairs TextureMapperLayer3DRenderingContext::sweepAndPruneXAxis(
+    const Vector<Layer>& layers)
+{
+    auto layerBounds = [](const Layer& layer) -> std::pair<float, float> {
+        return { layer.geometry.bounds().min().x(), layer.geometry.bounds().max().x() };
+    };
+
+    std::vector<size_t> indices(layers.size());
+    std::iota(indices.begin(), indices.end(), 0);
+
+    // Sort left to right along axis
+    std::sort(indices.begin(), indices.end(), [&layers, &layerBounds](size_t a, size_t b) {
+        auto [aMin, aMax] = layerBounds(layers[a]);
+        auto [bMin, bMax] = layerBounds(layers[b]);
+        return aMin < bMin;
     });
+
+    SweepAndPrunePairs potentialIntersectionPairs;
+    for (size_t i = 0; i < indices.size(); ++i) {
+        for (size_t j = i + 1; j < indices.size(); ++j) {
+            auto firstIndex = indices[i];
+            auto secondIndex = indices[j];
+
+            auto [iMin, iMax] = layerBounds(layers[firstIndex]);
+            auto [jMin, jMax] = layerBounds(layers[secondIndex]);
+            if (jMin >= iMax)
+                break; // No overlap possible
```

Comment:
How about adding bounding box intersection check here?

~~~c++
iMin = layers[firstIndex].geometry.bounds().min();
iMax = layers[firstIndex].geometry.bounds().max();
jMin = layers[secondIndex].geometry.bounds().min();
jMax = layers[secondIndex].geometry.bounds().max();
if (iMin.x() <= jMax.x())
    break;
if (iMin.y() >= jMax.y() || iMax.y() <= jMin.y())
    continue;
if (iMin.z() >= jMax.z() || iMax.z() <= jMin.z())
    continue;
~~~

Reply:
`if (iMin.x() <= jMax.x())` is wrong, but I think it's a typo and you probably meant `if (iMin.x() >= jMax.x())`.
But otherwise I think this should work. I cannot come up with any scenario why it wouldn't work.
I will run some more tests a bit later.

Reply:
This is now applied